src/routes/transactions/mappers/common/transaction-data.mapper.ts

```diff
@@ -0,0 +1,170 @@
+import { IContractsRepository } from '../../../../domain/contracts/contracts.repository.interface';
+import {
+  DELEGATE_OPERATION,
+  Operation,
+} from '../../../../domain/safe/entities/operation.entity';
+import { AddressInfoHelper } from '../../../common/address-info/address-info.helper';
+import { NULL_ADDRESS } from '../../../common/constants';
+import { AddressInfo } from '../../../common/entities/address-info.entity';
+import { isHex } from '../../../common/utils/utils';
+import { Contract } from '../../../contracts/entities/contract.entity';
+import { DataDecodedParameter } from '../../../data-decode/entities/data-decoded-parameter.entity';
+import { DataDecoded } from '../../../data-decode/entities/data-decoded.entity';
+import {
+  ADDRESS_PARAMETER_TYPE,
+  MULTI_SEND_METHOD_NAME,
+  TRANSACTIONS_PARAMETER_NAME,
+} from '../../constants';
+import { PreviewTransactionDto } from '../../entities/preview-transaction.dto.entity';
+import { TransactionData } from '../../entities/transaction-data.entity';
+import { DataDecodedParamHelper } from './data-decoded-param.helper';
+
+@Injectable()
+export class TransactionDataMapper {
+  constructor(
+    private readonly addressInfoHelper: AddressInfoHelper,
+    @Inject(IContractsRepository)
+    private readonly contractRepository: ContractsRepository,
+    private readonly dataDecodedParamHelper: DataDecodedParamHelper,
+  ) {}
+
+  async mapTransactionData(
+    chainId: string,
+    previewTransactionDto: PreviewTransactionDto,
+    dataDecoded: DataDecoded | null,
+  ): Promise<TransactionData> {
+    const toAddress = await this.addressInfoHelper.getOrDefault(
+      chainId,
+      previewTransactionDto.to,
+    );
+    const isTrustedDelegateCall = await this.isTrustedDelegateCall(
+      chainId,
+      previewTransactionDto.operation,
+      previewTransactionDto.to,
+      dataDecoded,
+    );
+    const addressInfoIndex = await this.buildAddressInfoIndex(
+      chainId,
+      dataDecoded,
+    );
+
+    return new TransactionData(
+      previewTransactionDto.data,
+      dataDecoded,
+      toAddress,
+      previewTransactionDto.value,
+      previewTransactionDto.operation,
+      isTrustedDelegateCall ?? null,
+      addressInfoIndex,
+    );
+  }
+
+  /**
+   * Determines if the transaction is a trusted DELEGATE call.
+   * @param chainId chain id
+   * @param operation transaction operation
+   * @param to transaction target address
+   * @param dataDecoded data decoded to check
+   * @returns null if the transaction operation is not DELEGATE.
+   * If the transaction operation is DELEGATE, and the target {@link Contract}
+   * is trusted, and the {@link DataDecoded} received contains a nested
+   * DELEGATE operation, then true is returned.
+   * Otherwise the function will return false.
+   */
+  async isTrustedDelegateCall(
+    chainId: string,
+    operation: Operation,
+    to: string,
+    dataDecoded: DataDecoded | null,
+  ): Promise<boolean | null> {
```

Comment:
We should reevaluate the return of this function. `isX` hints at a boolean return type or a future which provides a boolean. However we have 3 states being represented here: `true | false | null`. If we find that we really need 3 states I'd suggest representing those with an enum or similar.

This doesn't need to be tackled in this PR.

Reply:
Yeah exactly 👍 Even if the clients use `true | false | null` we can map this on the route level and work with a more meaningful domain representation.